content/en/blog/2025/12/v1-to-v2-transition/index.md

```diff
@@ -0,0 +1,368 @@
+aliases:
+  - /blog/2025/12/api-transition-v1-to-v2/
+---
+
+This blog article is going to walk you through the evolution of the RESTful API from v1 to v2, how they differ and what they have in common.  
+You'll see that we did some incremental improvements in how we organize sessions and the response structure.
+
+One major change to keep mind for Developers is that you need to create a project before you can store memories; before you were just shooting from the hip. 
+
+## v1 Flow
+
+Within the v1 API we organized the memories around sessions. When you added a memory you defined the following entities to capture important metadata around the message.  
+Let's use the following minimalistic request body towards `POST /v1/memories` as an example:
+```json
+{
+  "session": {
+    "group_id": "group_v1",
+    "user_id": [
+      "Christian",
+      "agent_v1"
+    ],
+    "session_id": "s-f62f517cf5fb997e97fa16"
+  },
+  "producer": "Christian",
+  "produced_for": "agent_v1",
+  "episode_content": "My name is Christian. I lived in Berlin and Hamburg.",
+  "episode_type": "message",
+  "metadata": {}
+}
+```
+
+The `session_id` identified a collection of connected messages. E.g. to indicate they where part of the same conversation.  
+All subsequent memories send with the same `session_id` were associated with the same conversation.
+
+Let's kick of the conversation by sending the following messages with the same `"session"` content to MemMachine v1.
+
+* My name is Christian. I lived in Berlin and Hamburg.
+* I like to play Minigolf on the weekends. Throughout the week I relax by playing tabletop games.
+* I am a container fanboy and like to tinker with different stacks on Kuberentes to keep me up to date.
+* My favorite books at the moment are crime novels.
+* When I am coding I like to listen to low tempo house music to get into 'the flow'.
+
+Once send we can query Neo4j to see how they are represented (CYPHER query: `MATCH (n) RETURN *`).
+
+![Figure 1. MemMachine v0.1 messages representation in Neo4j](v1-neo4j.png)
+
+As you can see all messages are neatly aligned. The nodes are organizes as follows:
+
+1. `Episodes` (purple): capture the complete content of the memory. 
+2. `Derivatives` (blue):  Chunk up individual sentences and prepare for LLM consumption by storing the embedding and prefix them with timestamp and producer.
+2. `EpisodeCluster` (orange): cluster `Derivates` (in case the Episode was rather long)
+
+The messages are also captured within a Pgvector (Postgres extension to store vector data) instance.
+
+![Figure 1. MemMachine v0.1 messages representation in pgweb](v1-pgweb.png)
+
+### Feature Extraction
+
+So far, so expected. MemMachine also extracts information from the `Derivatives` above and stores them in Profile Memory in the pgvector table `prof`:
+
+![Figure 1. MemMachine v0.1 messages representation in profile memory](v1-prof.png)
+
+You can query both memory types using the RESTful API:
+
+<table>
```

Comment:
I recommend using Markdown Tables rather than incorporating HTML in Markdown, which isn't recommended.